lib/Models/CatalogGroup.js

```diff
@@ -257,11 +232,41 @@ freezeObject(CatalogGroup.defaultSerializers);
  * @type {String[]}
  */
 CatalogGroup.defaultPropertiesForSharing = clone(CatalogMember.defaultPropertiesForSharing);
-CatalogGroup.defaultPropertiesForSharing.push('items');
 CatalogGroup.defaultPropertiesForSharing.push('isOpen');
+CatalogGroup.defaultPropertiesForSharing.push('items');
 
 freezeObject(CatalogGroup.defaultPropertiesForSharing);
 
+CatalogGroup.prototype.setupItemListeners = function() {
+    var itemsChangeListeners = {
+        added: function(item) {
+            item.parent = this;
+            item.allShareKeys.forEach(function (key) {
+                if (this.terria.catalog.shareKeyIndex[key]) {
+                    throw new RuntimeError('Duplicate shareKey: ' + key);
+                } else {
+                    this.terria.catalog.shareKeyIndex[key] = item;
+                }
+            }, this);
+        }.bind(this),
+        deleted: function(item) {
+            item.parent = undefined;
+
+            item.allShareKeys.forEach(function(key) {
+                this.terria.catalog.shareKeyIndex[key] = undefined;
```

Comment:
I think you need to do this before setting `item.parent = undefined`.  Otherwise the key you're removing won't be the same as the one you previously added.